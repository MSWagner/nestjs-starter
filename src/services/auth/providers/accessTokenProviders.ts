import { Connection, Repository } from "typeorm";
import { AccessToken } from "../../../entities/AccessToken.entity";

import CONFIG from "../../../config";

export const accessTokenProviders = [
    {
        provide: CONFIG.database.defaultAccessTokenRepoName,
        useFactory: (connection: Connection): Repository<AccessToken> => connection.getRepository(AccessToken),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
