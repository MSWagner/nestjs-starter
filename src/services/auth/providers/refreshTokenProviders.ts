import { Connection, Repository } from "typeorm";
import { RefreshToken } from "../../../entities/RefreshToken.entity";

import CONFIG from "../../../config";
import { AccessToken } from "../../../entities/AccessToken.entity";

export const refreshTokenProviders = [
    {
        provide: CONFIG.database.defaultRefreshTokenRepoName,
        useFactory: (connection: Connection): Repository<AccessToken> => connection.getRepository(RefreshToken),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
