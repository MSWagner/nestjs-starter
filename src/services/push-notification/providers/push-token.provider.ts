import { Connection, Repository } from "typeorm";
import { PushToken } from "../../../entities/PushToken.entity";

import CONFIG from "../../../config";

export const pushTokenProviders = [
    {
        provide: CONFIG.database.defaultPushTokenRepoName,
        useFactory: (connection: Connection): Repository<PushToken> => connection.getRepository(PushToken),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
