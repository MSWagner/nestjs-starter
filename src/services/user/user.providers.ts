import { Connection, Repository } from "typeorm";
import { User } from "../../entities/User.entity";

import CONFIG from "../../config";

export const userProviders = [
    {
        provide: CONFIG.database.defaultUserRepoName,
        useFactory: (connection: Connection): Repository<User> => connection.getRepository(User),
        inject: [CONFIG.database.defaultConnectionName]
    }
];
