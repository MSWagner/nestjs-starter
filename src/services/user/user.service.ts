import * as bcrypt from "bcrypt";
import { Repository } from "typeorm";
import { Injectable, Inject } from "@nestjs/common";

import { User } from "../../entities/User.entity";

import CONFIG from "../../config";

@Injectable()
export class UserService {
    private saltRounds = 10;

    constructor(
        @Inject(CONFIG.database.defaultUserRepoName)
        private readonly userRepository: Repository<User>
    ) {}

    private async getHash(password: string): Promise<string> {
        return bcrypt.hash(password, this.saltRounds);
    }

    async findOne(username: string): Promise<User | undefined> {
        return this.userRepository.findOne({ username }, { relations: ["userPermissions"] });
    }

    async createUser(username: string, password: string): Promise<User> {
        const user = new User();

        user.username = username;
        user.passwordHash = await this.getHash(password);

        return this.userRepository.save(user);
    }
}
