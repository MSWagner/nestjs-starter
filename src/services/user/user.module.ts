import { Module } from "@nestjs/common";
import { UserService } from "./user.service";
import { userProviders } from "./user.providers";

import { DatabaseModule } from "../database/database.module";

@Module({
    imports: [DatabaseModule],
    providers: [...userProviders, UserService],
    exports: [DatabaseModule, UserService, ...userProviders]
})
export class UserModule {}
