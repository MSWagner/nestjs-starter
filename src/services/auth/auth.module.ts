import { Module } from "@nestjs/common";

import { AuthService } from "./auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    exports: [AuthService]
})
export class AuthModule {}
