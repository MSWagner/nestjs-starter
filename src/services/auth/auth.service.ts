import * as bcrypt from "bcrypt";
import moment from "moment";
import * as _ from "lodash";

import { Injectable, Inject } from "@nestjs/common";
import { Repository } from "typeorm";

import { UserService } from "../user/user.service";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import { CONFIG } from "../../config";

export interface IToken {
    accessToken: AccessToken;
    refreshToken: RefreshToken;
}

@Injectable()
export class AuthService {
    constructor(
        @Inject(CONFIG.database.defaultAccessTokenRepoName)
        private readonly accessTokenRepository: Repository<AccessToken>,

        @Inject(CONFIG.database.defaultRefreshTokenRepoName)
        private readonly refreshTokenRepository: Repository<RefreshToken>,

        private readonly userService: UserService
    ) {}

    async generateToken(user: User): Promise<{ accessToken: AccessToken; refreshToken: RefreshToken }> {
        const newRefreshToken = new RefreshToken();
        newRefreshToken.validUntil =
            CONFIG.auth.refreshTokenValidityMS > 0
                ? moment().add(CONFIG.auth.refreshTokenValidityMS, "milliseconds").toDate()
                : null;
        newRefreshToken.user = user;

        const accessToken = await this.generateAuthToken(user);
        const refreshToken = await this.refreshTokenRepository.save(newRefreshToken);

        return {
            accessToken,
            refreshToken
        };
    }

    async generateAuthToken(user: User): Promise<AccessToken> {
        const newAccessToken = new AccessToken();
        newAccessToken.validUntil = moment().add(CONFIG.auth.tokenValidity, "milliseconds").toDate();
        newAccessToken.user = user;

        const accessToken = await this.accessTokenRepository.save(newAccessToken);

        return accessToken;
    }

    async validateUser(username: string, password: string): Promise<User> {
        const user = await this.userService.findOne(username);

        if (_.isNil(user)) {
            return null;
        }

        if (!user.isActive) {
            return null;
        }

        const isValid = await bcrypt.compare(password, user.passwordHash);

        return isValid ? user : null;
    }

    async validateToken(token: string): Promise<User> {
        const accessToken = await this.accessTokenRepository.findOne(token, { relations: ["user"] });

        if (_.isNil(accessToken)) {
            return null;
        }

        const isValid = moment().isSameOrBefore(accessToken.validUntil);

        if (!isValid) {
            await this.accessTokenRepository.delete(accessToken);
        }

        if (!accessToken.user.isActive) {
            return null;
        }

        return isValid ? accessToken.user : null;
    }

    async register(username: string, password: string): Promise<User> {
        return this.userService.createUser(username, password);
    }

    async login(user: User): Promise<IToken> {
        return this.generateToken(user);
    }

    async refreshAuthToken(token: string): Promise<IToken> {
        const refreshToken = await this.refreshTokenRepository.findOne(token, { relations: ["user"] });

        const isValid =
            !_.isNil(refreshToken) &&
            !_.isNil(refreshToken.user) &&
            (_.isNil(refreshToken.validUntil) || moment().isSameOrBefore(refreshToken.validUntil));

        if (!isValid && !_.isNil(refreshToken)) {
            await this.refreshTokenRepository.delete(refreshToken);
        }

        return !isValid
            ? null
            : {
                  accessToken: await this.generateAuthToken(refreshToken.user),
                  refreshToken
              };
    }
}
