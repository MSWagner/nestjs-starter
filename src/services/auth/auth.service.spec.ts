import { Test, TestingModule } from "@nestjs/testing";

import { TestModule } from "../../services/test/test.module";
import { TestService } from "../../services/test/test.service";

import { AuthService } from "./auth.service";
import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";

import { userProviders } from "../user/user.providers";
import { UserService } from "../user/user.service";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import * as fixtures from "../../services/test/fixtures";

describe("AuthService", () => {
    let service: AuthService;
    let testService: TestService;

    beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            providers: [AuthService, ...accessTokenProviders, ...refreshTokenProviders, UserService, ...userProviders],
            imports: [TestModule]
        }).compile();

        service = module.get<AuthService>(AuthService);
        testService = module.get<TestService>(TestService);

        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("should return new generated token for user1", async () => {
        const user1AccessTokenCount = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCount = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCount).toBe(1);
        expect(user1RefreshTokenCount).toBe(1);

        const user = await User.findOne({ where: { uid: fixtures.user1.uid } });
        const tokens = await service.generateToken(user);

        expect(tokens.accessToken).not.toBe(null);
        expect(tokens.refreshToken).not.toBe(null);

        const user1AccessTokenCountAfter = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCountAfter = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCountAfter).toBe(2);
        expect(user1RefreshTokenCountAfter).toBe(2);
    });

    it("should return validated user1 with user", async () => {
        const user = await service.validateUser(fixtures.user1.username, "testPassword");

        expect(user).not.toBe(null);

        const userWithoutDates = testService.replaceDates(user);
        expect(userWithoutDates).toMatchSnapshot("ValidatedUser1WithUser");
    });

    it("should return null for wrong user1 password", async () => {
        const user = await service.validateUser(fixtures.user1.username, "wrongPassword");

        expect(user).toBe(null);
    });

    it("should return null for wrong username", async () => {
        const user = await service.validateUser("wrongUsername", "testPassword");

        expect(user).toBe(null);
    });

    it("should return validated user1 with token", async () => {
        const user = await service.validateToken(fixtures.accessToken1.token);

        expect(user).not.toBe(null);

        const userWithoutDates = testService.replaceDates(user);
        expect(userWithoutDates).toMatchSnapshot("ValidatedUser1WithToken");
    });

    it("should return validated user3 as admin with token", async () => {
        const user = await service.validateToken(fixtures.accessTokenUser3Admin.token);

        expect(user).not.toBe(null);

        const userWithoutDates = testService.replaceDates(user);
        expect(userWithoutDates).toMatchSnapshot("ValidatedUser3AdminWithToken");
    });

    it("should return null for invalid token and delete it", async () => {
        const accessTokenCountBefore = await AccessToken.count({ token: fixtures.invalidAccessToken.token });
        expect(accessTokenCountBefore).toEqual(1);

        const user = await service.validateToken(fixtures.invalidAccessToken.token);

        expect(user).toBe(null);

        const accessTokenCountAfter = await AccessToken.count({ token: fixtures.invalidAccessToken.token });
        expect(accessTokenCountAfter).toEqual(0);
    });

    it("should register new user", async () => {
        const userCountBefore = await User.count();
        expect(userCountBefore).toBe(fixtures.fixtureTrees.User.length);

        const user = await service.register("newUsername", "password");

        expect(user).not.toBe(null);
        expect(user.username).toBe("newUsername");

        const userCountAfter = await User.count();
        expect(userCountAfter).toBe(fixtures.fixtureTrees.User.length + 1);
    });

    it("should throw error for username confict", async () => {
        const userCountBefore = await User.count();
        expect(userCountBefore).toBe(fixtures.fixtureTrees.User.length);

        try {
            const user = await service.register(fixtures.user1.username, "password");
            expect(user).toBe(null);
        } catch (err) {
            const errorMessage: string = err.message;

            expect(errorMessage).toInclude("duplicate key value violates unique constraint");
        }
    });

    it("should return new generated token for user1 login", async () => {
        const user1AccessTokenCount = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCount = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCount).toBe(1);
        expect(user1RefreshTokenCount).toBe(1);

        const user = await User.findOne({ where: { uid: fixtures.user1.uid } });
        const tokens = await service.login(user);

        expect(tokens.accessToken).not.toBe(null);
        expect(tokens.refreshToken).not.toBe(null);

        // To test the option with 0 as refreshTokenValidityMS
        expect(tokens.refreshToken.validUntil).toBe(null);

        const user1AccessTokenCountAfter = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCountAfter = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCountAfter).toBe(2);
        expect(user1RefreshTokenCountAfter).toBe(2);
    });

    it("should return new generated token with refreshToken", async () => {
        const user1AccessTokenCount = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCount = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCount).toBe(1);
        expect(user1RefreshTokenCount).toBe(1);

        const tokens = await service.refreshAuthToken(fixtures.refreshToken1.token);

        expect(tokens.accessToken).not.toBe(null);
        expect(tokens.refreshToken).not.toBe(null);

        const user1AccessTokenCountAfter = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCountAfter = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCountAfter).toBe(2);
        expect(user1RefreshTokenCountAfter).toBe(1);
    });

    it("should return null for not existing refreshToken", async () => {
        const user1AccessTokenCount = await AccessToken.count({ where: { user: fixtures.user1.uid } });
        const user1RefreshTokenCount = await RefreshToken.count({ where: { user: fixtures.user1.uid } });

        expect(user1AccessTokenCount).toBe(1);
        expect(user1RefreshTokenCount).toBe(1);

        const tokens = await service.refreshAuthToken("b393dafd-0492-417d-939f-1675c84b8e2d");

        expect(tokens).toBe(null);
    });

    it("should return null for invalid refreshToken", async () => {
        const refreshTokenCountBefore = await RefreshToken.count({ token: fixtures.invalidRefreshToken.token });
        expect(refreshTokenCountBefore).toBe(1);

        const tokens = await service.refreshAuthToken(fixtures.invalidRefreshToken.token);

        expect(tokens).toBe(null);

        const refreshTokenCountAfter = await RefreshToken.count({ token: fixtures.invalidRefreshToken.token });
        expect(refreshTokenCountAfter).toBe(0);
    });

    it("should return null for inactive user validation", async () => {
        const user = await service.validateUser(fixtures.userInActive.username, "testPassword");

        expect(user).toBe(null);
    });
});
