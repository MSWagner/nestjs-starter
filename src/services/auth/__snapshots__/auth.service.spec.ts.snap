// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService should return validated user1 with token: ValidatedUser1WithToken 1`] = `
User {
  "createdAt": "createdAt",
  "isActive": true,
  "passwordHash": "$2b$10$n8McQmLm6U/RXlRb9HxNEOgurQecAokxI2oCoumlSOlOj4M.aU59K",
  "pushTokens": [
    PushToken {
      "createdAt": "createdAt",
      "token": "user1PushTokenFixtures",
      "uid": "733d80fd-d861-4638-99c8-fcc5c5e93963",
      "updatedAt": "updatedAt",
    },
  ],
  "uid": "0961866f-ff60-4219-973c-109eecfa52bd",
  "updatedAt": "updatedAt",
  "userPermissions": [
    UserPermission {
      "createdAt": "createdAt",
      "permission": Permission {
        "createdAt": "createdAt",
        "scope": "user",
        "uid": "d7316b3e-44de-42b8-bc9b-bbf190cd98fa",
        "updatedAt": "updatedAt",
      },
      "updatedAt": "updatedAt",
    },
  ],
  "username": "user1@email.at",
}
`;

exports[`AuthService should return validated user1 with user: ValidatedUser1WithUser 1`] = `
User {
  "createdAt": "createdAt",
  "isActive": true,
  "passwordHash": "$2b$10$n8McQmLm6U/RXlRb9HxNEOgurQecAokxI2oCoumlSOlOj4M.aU59K",
  "pushTokens": [
    PushToken {
      "createdAt": "createdAt",
      "token": "user1PushTokenFixtures",
      "uid": "733d80fd-d861-4638-99c8-fcc5c5e93963",
      "updatedAt": "updatedAt",
    },
  ],
  "uid": "0961866f-ff60-4219-973c-109eecfa52bd",
  "updatedAt": "updatedAt",
  "userPermissions": [
    UserPermission {
      "createdAt": "createdAt",
      "permission": Permission {
        "createdAt": "createdAt",
        "scope": "user",
        "uid": "d7316b3e-44de-42b8-bc9b-bbf190cd98fa",
        "updatedAt": "updatedAt",
      },
      "updatedAt": "updatedAt",
    },
  ],
  "username": "user1@email.at",
}
`;

exports[`AuthService should return validated user3 as admin with token: ValidatedUser3AdminWithToken 1`] = `
User {
  "createdAt": "createdAt",
  "isActive": true,
  "passwordHash": null,
  "pushTokens": [
    PushToken {
      "createdAt": "createdAt",
      "token": "user3PushTokenFixtures",
      "uid": "ff5fce38-ca56-4e08-be2c-7af095266bf6",
      "updatedAt": "updatedAt",
    },
  ],
  "uid": "435c0a53-9e6b-488a-9d1c-582b108ed3e7",
  "updatedAt": "updatedAt",
  "userPermissions": [
    UserPermission {
      "createdAt": "createdAt",
      "permission": Permission {
        "createdAt": "createdAt",
        "scope": "admin",
        "uid": "ebf943cd-cd67-4720-bd40-6cf3c25e4054",
        "updatedAt": "updatedAt",
      },
      "updatedAt": "updatedAt",
    },
  ],
  "username": "user3@email.com",
}
`;
